pkg/pdh/win_perf_counters.go

```diff
@@ -0,0 +1,620 @@
+					counterPath,
+					computer,
+					objectName,
+					instance,
+					counterName,
+					measurement,
+					includeTotal,
+					useRawValue,
+				)
+			}
+
+			if instance == "_Total" && origInstance == "*" && !includeTotal {
+				continue
+			}
+
+			hostCounter.counters = append(hostCounter.counters, newItem)
+		}
+	} else {
+		newItem := newCounter(
+			counterHandle,
+			counterPath,
+			computer,
+			objectName,
+			instance,
+			counterName,
+			measurement,
+			includeTotal,
+			useRawValue,
+		)
+		hostCounter.counters = append(hostCounter.counters, newItem)
+	}
+
+	return nil
+}
+
+const emptyInstance = "------"
+
+func formatPath(computer, objectName, instance, counter string) string {
+	path := ""
+	if instance == emptyInstance {
+		path = fmt.Sprintf(`\%s\%s`, objectName, counter)
+	} else {
+		path = fmt.Sprintf(`\%s(%s)\%s`, objectName, instance, counter)
+	}
+	if computer != "" && computer != "localhost" {
+		path = fmt.Sprintf(`\\%s%s`, computer, path)
+	}
+	return path
+}
+
+func (m *WinPerfCounters) ParseConfig() error {
+	var counterPath string
+	m.queryCreator = &PerformanceQueryCreatorImpl{}
+
+	if m.MaxBufferSize == 0 {
+		m.MaxBufferSize = int64(defaultMaxBufferSize)
+	}
+
+	if len(m.Sources) == 0 {
+		m.Sources = []string{"localhost"}
+	}
+
+	if len(m.Object) <= 0 {
+		err := errors.New("no performance objects configured")
+		return err
+	}
+
+	for _, PerfObject := range m.Object {
+		computers := PerfObject.Sources
+		if len(computers) == 0 {
+			computers = m.Sources
+		}
+		for _, computer := range computers {
+			if computer == "" {
+				// localhost as a computer name in counter path doesn't work
+				computer = "localhost"
+			}
+			for _, counter := range PerfObject.Counters {
+				if len(PerfObject.Instances) == 0 {
+					_ = level.Warn(m.Log).Log("msg", fmt.Sprintf("Missing 'Instances' param for object %q", PerfObject.ObjectName))
+				}
+				for _, instance := range PerfObject.Instances {
+					objectName := PerfObject.ObjectName
+					counterPath = formatPath(computer, objectName, instance, counter)
+
+					err := m.AddItem(counterPath, computer, objectName, instance, counter,
+						PerfObject.Measurement, PerfObject.IncludeTotal, PerfObject.UseRawValues)
+					if err != nil {
+						if PerfObject.FailOnMissing || PerfObject.WarnOnMissing {
+							_ = level.Error(m.Log).Log("msg", fmt.Sprintf("Invalid counterPath %q: %s", counterPath, err.Error()))
+						}
+						if PerfObject.FailOnMissing {
+							return err
+						}
+					}
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func (m *WinPerfCounters) checkError(err error) error {
+	var pdhErr *PdhError
+	if errors.As(err, &pdhErr) {
+		for _, ignoredErrors := range m.IgnoredErrors {
+			if PDHErrors[pdhErr.ErrorCode] == ignoredErrors {
+				return nil
+			}
+		}
+
+		return err
+	}
+	return err
+}
+
+func (m *WinPerfCounters) GetInfo() (map[string]CounterInfos, error) {
+	var wg sync.WaitGroup
+	info := map[string]CounterInfos{}
+
+	// iterate over computers
```

Comment:
Technically, it's possible to fetch perfdata from remote machines.

Not sure, if we ever want to support this use-case. However I would avoid to change code here. Otherwise it's every hard to copy and backport fixes from upstream.

Reply:
Sure, keeping the code unchanged is a good reason. Perhaps add a comment with the context you've provided somewhere in or above that function?

Just seeking to avoid confusion in the future whenever the code needs to be read.

Reply:
I added a comment on the package level, is that sufficent?

Reply:
Yep that's great :+1: